### PR TITLE
#177728140 add remittance to transaction model

### PR DESCRIPTION
**Pull request by tobieeniafe**
#### What does this PR do?
Add remittance to transaction model

#### Description of Task to be completed?
Add remittance to transaction model

#### How should this be manually tested?
`PUT /api/v1/transaction/remittance`

#### What are the relevant pivotal tracker stories?
#177728140

### CODE REVIEW

File: server/services/transaction.service.js [hunk 1]
```diff
@@ -270,3 +272,34 @@ export const getOneTransaction = async (transactionId, user) => {
...
+export const addRemittance = async (remittanceInfo) => {
+  const transaction = await getTransactionById(remittanceInfo.transactionId).catch((error) => {
+    throw new ErrorHandler(httpStatus.INTERNAL_SERVER_ERROR, 'Internal Server Error', error);
+  });
+
+  if (!transaction) {
+    throw new ErrorHandler(httpStatus.NOT_FOUND, 'Invalid transaction');
+  }
+
+  const vendor = await getUserById(transaction.vendorId).catch((error) => {
```
Comment (andela-hpopoola) [inline server/services/transaction.service.js:285]:
```suggestion
  const user = await getUserById(transaction.vendorId).catch((error) => {
```

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for working on this. I think this looks good to me pending the time you work on the test. 

Also, can we ensure we update the `getAllTransaction` tests to ensure that remittance is not returned when a user is getting transactions (we don't want the user to know about remittance) but this should be available when an admin/vendor get all transactions?

File: server/services/transaction.service.js [hunk 2]
```diff
@@ -122,6 +122,10 @@ export const getAllTransactions = async (user, page = 1, limit = 10) => {
 
   if (user.role === USER_ROLE.USER) {
     transactionOptions.unshift({ $match: { userId: ObjectId(user._id) } });
+    transactionOptions[7].$project = {
```
Comment (andela-hpopoola) [inline server/services/transaction.service.js:125]: This looks like a hacky way to exclude transaction info. A better alternative is to create it as a function that returns `EXCLUDED_TRANSACTION_INFO` if the current user has a user role else it returns null or an empty object.

Something like this:

```js
{
      $project: {
        ...NON_PROJECTED_USER_INFO('userInfo'),
        ...NON_PROJECTED_USER_INFO('vendorInfo'),
        ...EXCLUDED_PROPERTY_INFO,
        ...getExcludedTransactionInfo(userRole),
      },
    },
```

File: server/services/transaction.service.js [hunk 3]
```diff
@@ -177,6 +177,10 @@ export const getUserTransactionsByProperty = async (propertyId, user, page = 1,
 
   if (user.role === USER_ROLE.USER) {
     transactionOptions.unshift({ $match: { userId: ObjectId(user._id) } });
+    transactionOptions[6].$project = {
```
Comment (andela-hpopoola) [inline server/services/transaction.service.js:180]: Same comment as above.

File: server/services/transaction.service.js [hunk 4]
```diff
@@ -267,6 +265,10 @@ export const getOneTransaction = async (transactionId, user) => {
 
   if (user.role === USER_ROLE.USER) {
     transactionOptions.unshift({ $match: { userId: ObjectId(user._id) } });
+    transactionOptions[8].$project = {
```
Comment (andela-hpopoola) [inline server/services/transaction.service.js:268]: Still same comment as above

File: test/controllers/transaction.controller.test.js [hunk 5]
```diff
@@ -207,6 +221,7 @@ describe('Transaction Controller', () => {
               expect(res.body.result[0].offerId).to.be.eql(
                 vendor2Transactions[0].offerId.toString(),
               );
+              expect(res.body.result[0]).to.have.property('remittance');
```
Comment (andela-hpopoola) [inline test/controllers/transaction.controller.test.js:224]: Let's test the values of remittance returned instead of checking if the `remittance` property exists here. It is better to check for the amount, by, percentage e.t.c since they are available in the object.

File: test/controllers/transaction.controller.test.js [hunk 6]
```diff
@@ -776,4 +843,103 @@ describe('Transaction Controller', () => {
...
+        propertyId: property._id,
+        offerId: offer._id,
+        vendorId: vendorUser._id,
+        addedBy: adminUser._id,
+        updatedBy: adminUser._id,
+      },
+      { generateId: true },
+    );
+
+    const data = {
```
Comment (andela-hpopoola) [inline test/controllers/transaction.controller.test.js:863]: Can we add a test with the `percentage` key defined in `data` since percentage is optional.

File: server/controllers/transaction.controllers.js [hunk 7]
```diff
@@ -70,6 +71,16 @@ const TransactionController = {
       })
       .catch((error) => next(error));
   },
+
+  addRemittance(req, res, next) {
+    const remittanceInfo = req.locals;
+    const adminId = req.user._id;
+    addRemittance({ ...remittanceInfo, adminId })
```
Comment (andela-hpopoola) [inline server/controllers/transaction.controllers.js:78]: A mail should be sent to the vendor that he has been paid.

File: test/controllers/transaction.controller.test.js [hunk 8]
```diff
@@ -776,4 +843,103 @@ describe('Transaction Controller', () => {
...
+      transactionId: transaction._id,
+      amount: 1_000_000,
+      date: '2021-02-12',
+    };
+
+    beforeEach(async () => {
+      await Transaction.create(transaction);
+    });
+
+    context('with valid token and data', () => {
```
Comment (andela-hpopoola) [inline test/controllers/transaction.controller.test.js:873]: Can we also create a user with a different remittance (from 5) to ensure that remittance from the user model works.

File: server/schemas/transaction.schema.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import Joi from '@hapi/joi';
+import { requiredObjectId, requiredNumber, requiredDate, optionalNumber } from './helper.schema';
+
+const addRemittanceSchema = Joi.object({
+  transactionId: requiredObjectId('Transaction Id'),
+  amount: requiredNumber('Amount'),
```
Comment (andela-hpopoola) [inline server/schemas/transaction.schema.js:6]: On a closer look, we shouldn't ask for the amount from the endpoint, we should rather calculate this from the supplied percentage. That way the amount will always be consistent with the percentage and would fix the bug of remitting more than the amount paid for the transaction.

File: server/models/user.model.js [hunk 10]
```diff
@@ -299,6 +299,10 @@ const UserSchema = new mongoose.Schema(
           updatedAt: Date,
         },
       ],
+      remittancePercentage: {
+        type: Number,
+        default: 5,
```
Comment (andela-hpopoola) [inline server/models/user.model.js:304]: Don't forget, you need to confirm this default from Nnamdi.

File: server/services/transaction.service.js [hunk 11]
```diff
@@ -265,8 +265,43 @@ export const getOneTransaction = async (transactionId, user) => {
...
+  const user = await getUserById(transaction.vendorId).catch((error) => {
+    throw new ErrorHandler(httpStatus.INTERNAL_SERVER_ERROR, 'Internal Server Error', error);
+  });
+
+  try {
+    return Transaction.findByIdAndUpdate(
+      transaction.id,
+      {
+        $set: {
+          'remittance.amount': remittanceInfo.amount,
```
Comment (andela-hpopoola) [inline server/services/transaction.service.js:296]: This should be calculated from the percentage and rounded to the nearest whole number for consistency sake

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for making the changes. I left some more comments on ensuring that we cover more edge cases concerning this endpoint. Thank you and Nice work.

File: test/controllers/transaction.controller.test.js [hunk 12]
```diff
@@ -105,6 +107,12 @@ describe('Transaction Controller', () => {
         vendorId: vendorUser._id,
         addedBy: adminUser._id,
         updatedBy: adminUser._id,
+        remittance: {
```
Comment (andela-hpopoola) [inline test/controllers/transaction.controller.test.js:110]: Lastly, In another PR (maybe next week), we would need an endpoint to enable an admin to update a vendor's remittance percentage.

File: mailer/index.js [hunk 13]
```diff
@@ -79,6 +79,11 @@ const EMAIL_CONTENT = {
     contentBottom: `Click on the link below to accept their invitation`,
     buttonText: 'Join BALLERS',
   },
+  REMITTANCE_PAID: {
+    subject: 'Remittance Paid',
```
Comment (andela-hpopoola) [inline mailer/index.js:83]: This will surely not be remittance paid. Please work with content team to get a nice email to send out on this.

File: server/helpers/projectedSchemaInfo.js [hunk 14]
```diff
@@ -73,6 +75,9 @@ export const NON_PROJECTED_USER_INFO = (infoType) => {
   };
 };
 
-export const EXCLUDED_TRANSACTION_INFO = {
-  remittance: 0,
+export const GET_EXCLUDED_TRANSACTION_INFO = (userRole) => {
```
Comment (andela-hpopoola) [inline server/helpers/projectedSchemaInfo.js:78]: If this is a function, then the naming convention shouldn't be in ALL CAPS.

```js
export const getExcludedTransactionInfo = (userRole) => {
```

File: server/helpers/projectedSchemaInfo.js [hunk 15]
```diff
@@ -73,6 +75,9 @@ export const NON_PROJECTED_USER_INFO = (infoType) => {
...
+  if (userRole === USER_ROLE.USER) {
```
Comment (andela-hpopoola) [inline server/helpers/projectedSchemaInfo.js:79]: This can be written as one liner:

```js
export const getExcludedTransactionInfo = (userRole) => userRole === USER_ROLE.USER ? { remittance: 0 } : {};
````

File: server/schemas/transaction.schema.js [hunk 16]
```diff
@@ -1,9 +1,8 @@
 import Joi from '@hapi/joi';
-import { requiredObjectId, requiredNumber, requiredDate, optionalNumber } from './helper.schema';
+import { requiredObjectId, requiredDate, optionalNumber } from './helper.schema';
 
 const addRemittanceSchema = Joi.object({
   transactionId: requiredObjectId('Transaction Id'),
-  amount: requiredNumber('Amount'),
   date: requiredDate('Date'),
   percentage: optionalNumber('Percentage'),
```
Comment (andela-hpopoola) [inline server/schemas/transaction.schema.js:7]: Can we create an optionalPercentage function to cater for this so that we don't accept values over 100% or below 0. Lets also add test for less than 0 and greater than 100 🙏🏽

File: server/routes/transaction.routes.js [hunk 17]
```diff
@@ -142,4 +145,35 @@ router.get(
...
+ *      description: Adds remittance for a specific transaction
+ *     responses:
+ *      '200':
+ *        description: Remittance added
+ *      '400':
+ *        description: Error adding remittance
+ *      '500':
+ *       description: Internal server error
+ */
+router.put(
```
Comment (andela-hpopoola) [inline server/routes/transaction.routes.js:171]: Can we change this to `post` since we are using the endpoint to add a new remittance 🙏🏽 . This makes the endpoint consistent even if we need to move the remittance model into its own model. It also saves us the stress if we wish to create an endpoint to update remittance. 

I know it is confusing but I feel post is the right thing to use here even though we are updating the transaction model.

```suggestion
router.post(
```

**Review by andela-hpopoola — APPROVED** [review]
Thank you for working on this. I left a few comments to perfect this. Feel free to merge once you fix them. Thank you.

File: server/models/user.model.js [hunk 18]
```diff
@@ -299,6 +299,10 @@ const UserSchema = new mongoose.Schema(
           updatedAt: Date,
         },
       ],
+      remittancePercentage: {
```
Comment (andela-hpopoola) [inline server/models/user.model.js:302]: I believe this is sensitive information. Can we ensure that this is also not leaked to users too. Feel free to implement this in a new PR if it is tricky.
